### PR TITLE
feat: Support customizable deployment strategy for RawDeployment mode. Fixes #3452

### LINKED ISSUE #3452: Ability to change parameters of of the default `Deployment` created from `ServingRuntime` and `InferenceService`
/kind feature

**Describe the solution you'd like**
I'd like to have the ability to modify some of the `Deployment.spec` for the predictor deployment created from  `ServingRuntime` and `InferenceService`. These parameters include `strategy` for adjusting rolling update strategy, `template.spec.securityContext` for adjusting the security context to name a few. The ideal solution would be to override any `Deployment.spec` inside the `ServingRuntime` and `InferenceService`.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]
Currently, the deployment created by `ServingRuntime` and `InferenceService` use the default Kubernetes deployment values. 

**Links to the design documents:**
[Optional, start with the short-form [RFC template](https://docs.google.com/document/d/1UcBeLfZ_JMGpVrPJmYtEIVH_9Y4U3AEKQdq_IKuOMrU) to outline your ideas and get early feedback.]
[Required, use the longer-form [design doc template](https://docs.google.com/document/d/1Mtoui_PP2a9N59NjYHnsvrdJ8t2iKFwIJAx1zRO_I1c) to specify and discuss your design in more detail]

### PR DESCRIPTION
**Pull request by terrytangyuan**
Fixes #3452. Supersedes https://github.com/kserve/kserve/pull/3479

### CODE REVIEW

**Comment by spolti** [comment]
Maybe we could migrate the formatting changes to another PR?

File: pkg/apis/serving/v1beta1/component.go [hunk 1]
```diff
@@ -106,6 +107,10 @@ type ComponentExtensionSpec struct {
 	// More info: http://kubernetes.io/docs/user-guide/annotations
 	// +optional
 	Annotations map[string]string `json:"annotations,omitempty"`
+
+	// The deployment strategy to use to replace existing pods with new ones.
+	// +optional
+	DeploymentStrategy *appsv1.DeploymentStrategy `json:"deploymentStrategy,omitempty"`
```
Comment (spolti) [inline pkg/apis/serving/v1beta1/component.go:113]: defaults to?

**Comment by terrytangyuan** [comment]
> Maybe we could migrate the formatting changes to another PR?

The changes that get reformatted are introduced in this PR

File: pkg/controller/v1beta1/inferenceservice/rawkube_controller_test.go [hunk 2]
```diff
@@ -503,6 +503,420 @@ var _ = Describe("v1beta1 inference service controller", func() {
 			}
 			Expect(actualHPA.Spec).To(gomega.Equal(expectedHPA.Spec))
 		})
+		It("Should have ingress/service/deployment/hpa created with DeploymentStrategy", func() {
+			By("By creating a new InferenceService with DeploymentStrategy in PredictorSpec")
```
Comment (spolti) [inline pkg/controller/v1beta1/inferenceservice/rawkube_controller_test.go:507]: can't be this test broken into smaller `By` steps?

**Comment by spolti** [comment]
> > Maybe we could migrate the formatting changes to another PR?
> 
> The changes that get reformatted are introduced in this PR

even the jupyter notebook ones?

**Comment by terrytangyuan** [comment]
> > > Maybe we could migrate the formatting changes to another PR?
> 
> > 
> 
> > The changes that get reformatted are introduced in this PR
> 
> 
> 
> even the jupyter notebook ones?

Oh didn't notice those. I'll remove them.

File: pkg/controller/v1beta1/inferenceservice/rawkube_controller_test.go (hunk 2 above)
Comment (terrytangyuan) [reply pkg/controller/v1beta1/inferenceservice/rawkube_controller_test.go:507]: Nope. It would require refactoring existing tests

File: pkg/apis/serving/v1beta1/component.go (hunk 1 above)
Comment (terrytangyuan) [reply pkg/apis/serving/v1beta1/component.go:113]: It's optional and defaults to whatever appsv1.DeploymentStrategy has.

**Comment by terrytangyuan** [comment]
@spolti Actually these notebook changes will have to be part of this PR to pass the CI build.

**Review by spolti — APPROVED** [review]

File: pkg/controller/v1beta1/inferenceservice/reconcilers/deployment/deployment_reconciler.go [hunk 3]
```diff
@@ -77,6 +77,9 @@ func createRawDeployment(componentMeta metav1.ObjectMeta,
 			},
 		},
 	}
+	if componentExt.DeploymentStrategy != nil {
+		deployment.Spec.Strategy = *componentExt.DeploymentStrategy
```
Comment (Jooho) [inline pkg/controller/v1beta1/inferenceservice/reconcilers/deployment/deployment_reconciler.go:81]: From my understanding, rawDeployment is using HPA as the default autoscaler. This means that HPA will overwrite deployment replicas. So I guess hpa and this controller will continuously try to update this deployment, what do you think?

Comment (terrytangyuan) [reply pkg/controller/v1beta1/inferenceservice/reconcilers/deployment/deployment_reconciler.go:81]: It should be fine since HPA will only update the replicas field. Here we are only modifying the deployment strategy so it won't be overwritten.

Comment (Jooho) [reply pkg/controller/v1beta1/inferenceservice/reconcilers/deployment/deployment_reconciler.go:81]: Ok it only changed strategy so this source is ok.

File: pkg/apis/serving/v1beta1/component.go (hunk 1 above)
Comment (yuzisun) [inline pkg/apis/serving/v1beta1/component.go:113]: What happens when this field is set for the Knative mode?

File: pkg/apis/serving/v1beta1/component.go [hunk 4]
```diff
@@ -106,6 +107,10 @@ type ComponentExtensionSpec struct {
 	// More info: http://kubernetes.io/docs/user-guide/annotations
 	// +optional
 	Annotations map[string]string `json:"annotations,omitempty"`
+
+	// The deployment strategy to use to replace existing pods with new ones.
```
Comment (yuzisun) [inline pkg/apis/serving/v1beta1/component.go]: Can we add a note that this field is only applicable for raw deployment mode

File: pkg/apis/serving/v1beta1/component.go (hunk 1 above)
Comment (terrytangyuan) [reply pkg/apis/serving/v1beta1/component.go:113]: This is only used in `NewDeploymentReconciler.createRawDeployment` so it won't affect knative mode.

File: pkg/apis/serving/v1beta1/component.go (hunk 4 above)
Comment (terrytangyuan) [reply pkg/apis/serving/v1beta1/component.go]: Done

File: docs/samples/client/kfserving_sdk_v1beta1_sample.ipynb [hunk 5]
```diff
@@ -22,7 +22,7 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "from kubernetes import client \n",
+    "from kubernetes import client\n",
```
Comment (yuzisun) [inline docs/samples/client/kfserving_sdk_v1beta1_sample.ipynb:25]: Next time let's separate these changes into a separate PR which can make PR review easier

File: pkg/apis/serving/v1beta1/component.go (hunk 1 above)
Comment (yuzisun) [reply pkg/apis/serving/v1beta1/component.go:113]: the field is still exposed on the isvc which user can set for the knative mode, but I guess knative probably validate this.

File: docs/samples/client/kfserving_sdk_v1beta1_sample.ipynb (hunk 5 above)
Comment (terrytangyuan) [reply docs/samples/client/kfserving_sdk_v1beta1_sample.ipynb:25]: Somehow I cannot avoid it (haven't figured out why yet). These changes happen automatically by black+pre-commit and the build would fail without including the changes.

File: pkg/apis/serving/v1beta1/component.go (hunk 1 above)
Comment (yuzisun) [reply pkg/apis/serving/v1beta1/component.go:113]: I think we need to add a validation that this field is not supported for knative otherwise user will expect this to work but knative performs blue/green deployment.

**Comment by bmopuri** [comment]
@terrytangyuan   Thank you for making these changes.  I see these deployment strategy targets inference services in raw deployment mode.  Starting with kserve v12 we support raw deployment for inference graphs as well.

Could you please extend this change to inference graph raw deployment as well?

File: pkg/apis/serving/v1beta1/component.go (hunk 1 above)
Comment (terrytangyuan) [reply pkg/apis/serving/v1beta1/component.go:113]: Added

**Comment by terrytangyuan** [comment]
>Could you please extend this change to inference graph raw deployment as well?

I am trying to narrow down the scope of this PR. Extending this to IG would be a separate PR.
